File: educative/js/find_word_concatenation.js
function find_word_concatenation(str, arr) {
  if (arr.length === 0 || arr[0].length === 0) {
    return [];
  }
  const results = [];
  const wordFrequency = {};
  const eachWordLength = arr[0].length;
  const wordscount = arr.length;

  arr.map((word) => {
    if (!(word in wordFrequency)) {
      wordFrequency[word] = 0;
    }
    wordFrequency[word] += 1;
  });

  for (let i = 0; i < str.length - wordscount * eachWordLength + 1; i++) {
    const wordsSeen = {};
    for (let j = 0; j < wordscount; j++) {
      wordIndex = i + j * eachWordLength;

      word = str.substring(wordIndex, wordIndex + eachWordLength);
      if (!(word in wordFrequency)) {
        break;
      }
      if (!(word in wordsSeen)) {
        wordsSeen[word] = 0;
      }

      wordsSeen[word]++;

      if (wordsSeen[word] > wordFrequency[word] || 0) {
        console.log(j, "not here");
        break;
      }

      if (j + 1 === wordscount) {
        results.push(i);
      }
    }
  }
  return results;
}

console.log(find_word_concatenation("catfoxcat", ["cat", "fox"]));
// console.log(find_word_concatenation("catcatfoxfox", ["cat", "fox"]));
// console.log(find_word_concatenation("cmapcapmap", ["cap", "map"]));

File: educative/cpp/find_word_concatenation.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>
using namespace std;

vector<int> find_word_concatenation(const string &str, const vector<string> &words)
{
  vector<int> resultIndices;
  int wordsCount = words.size();
  int wordLength = words[0].length();
  unordered_map<string, int> wordFrequencyMap;
  for (auto word : words)
  {
    wordFrequencyMap[word]++;
  }
  for (int i = 0; i < (str.length() - wordsCount * wordLength + 1); i++)
  {
    unordered_map<string, int> seen;
    for (int j = 0; j < words.size(); j++)
    {
      int wordindex = i + j * wordLength;
      string word = str.substr(wordindex, wordLength);
      if (wordFrequencyMap.find(word) != wordFrequencyMap.end())
      {
        break;
      }
      seen[word]++;
      if (seen[word] > wordFrequencyMap[word])
      {
        break;
      }
      if (j + 1 == wordsCount)
      {
        resultIndices.push_back(i);
      }
    }
  }

  return resultIndices;
}

int main()
{
  vector<int> chars = find_word_concatenation("catfoxcat", vector<string>{"cat", "fox"});
}